community/kernel/src/main/java/org/neo4j/kernel/impl/index/schema/fusion/InstanceSelector.java

```diff
@@ -51,17 +58,137 @@
         this.factory = factory;
     }
 
+    /**
+     * Returns the instance at the given slot. Instantiating it if it hasn't already been instantiated.
+     *
+     * @param slot slot number to return instance for.
+     * @return the instance at the given {@code slot}.
+     */
     T select( int slot )
     {
         if ( instances[slot] == null )
         {
+            assertOpen();
             instances[slot] = factory.apply( slot );
         }
         return instances[slot];
     }
 
+    private void assertOpen()
+    {
+        if ( closed )
+        {
+            throw new IllegalStateException( "This selector has been closed" );
+        }
+    }
+
+    /**
+     * Returns the instance at the given slot. If the instance at the given {@code slot} hasn't been instantiated yet, {@code null} is returned.
+     *
+     * @param slot slot number to return instance for.
+     * @return the instance at the given {@code slot}, or {@code null}.
+     */
     T getIfInstantiated( int slot )
     {
         return instances[slot];
     }
+
+    /**
+     * Convenience method typically for calling a method on each of the instances, a method that returns another type of instance.
+     * Even called on instances that haven't been instantiated yet. All those created instances are put into the provided array and returned.
+     *
+     * @param target array to put the created instances into, also returned.
+     * @param converter {@link ThrowingFunction} which converts from the source to target instance.
+     * @param <R> type of returned instance.
+     * @param <E> type of exception that converter may throw.
+     * @return the target array which was passed in, now populated.
+     * @throws E exception from converter.
+     */
+    <R,E extends Exception> R[] instancesAs( R[] target, ThrowingFunction<T,R,E> converter ) throws E
```

Comment:
Moving all of those methods here from `FusionIndexBase` made things so much clearer! 🎉